File: clients/fsharp/Model/BuildingAction.fs
#nowarn "0058"

namespace SpbAiChamp
namespace SpbAiChamp.Model

open SpbAiChamp

/// Building order
type BuildingAction = {
    /// Id of the planet where the action needs to be performed
    Planet: int;
    /// Type of a building to build. If absent, current building will be destroyed
    BuildingType: option<Model.BuildingType>;
} with

    /// Write BuildingAction to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Planet
        match this.BuildingType with
            | Some value ->
                writer.Write true
                writer.Write (int value)
            | None -> writer.Write false
        ()

    /// Read BuildingAction from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Planet = reader.ReadInt32()
        BuildingType = match reader.ReadBoolean() with
                           | true -> Some(reader.ReadInt32() |> enum)
                           | false -> None
    }

File: clients/rust/src/model/building_action.rs
use super::*;

/// Building order
#[derive(Clone, Debug)]
pub struct BuildingAction {
    /// Id of the planet where the action needs to be performed
    pub planet: i32,
    /// Type of a building to build. If absent, current building will be destroyed
    pub building_type: Option<model::BuildingType>,
}

impl trans::Trans for BuildingAction {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.planet.write_to(writer)?;
        self.building_type.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let planet: i32 = trans::Trans::read_from(reader)?;
        let building_type: Option<model::BuildingType> = trans::Trans::read_from(reader)?;
        Ok(Self {
            planet,
            building_type,
        })
    }
}